### PR TITLE
FCREPO-3704: Close internally used ClosableHttpClient instance

### PR DESCRIPTION
**Pull request by tomcbe**
**Close internally used ClosableHttpClient instance**
* * *

**JIRA Ticket**: https://fedora-repository.atlassian.net/browse/FCREPO-3704

# What does this Pull Request do?
Properly close the internally used instance of CloseableHttpClient when FcrepoClient is closed.

# What's new?
* Implements the Closeable interace for FcrepoClient
* Closes the internally used instance of CloseableHttpClient in FcrepoClient in the newly added close() method.
* Refactors tests and integration tests to properly close resources.

# How should this be tested?
* Existing tests demonstrate, that this change as no effect on usage of FcrepoClient

# Additional Notes:
* Currently there is no test that explicitly tests, if the CloseableHttpClient has been closed

# Interested parties
@dbernstein

### CODE REVIEW

**Comment by tomcbe** [comment]
@dbernstein This is my current progress regarding FCREPO-3704.

I still have some open questions:
* How do we communicate to users that they should close any instances of FcrepoClient, when no longer needed?
* We currently use reflection (see method `TestUtils.setField`) to set the private field httpClient in FcrepoClient in some tests. But this always leaves a unmanaged instance of ClosableHttpClient that is never closed during tests. Do you see a proper way to tackle this problem?

**Comment by dbernstein** [comment]
@tomcbe :  sorry - this PR fell off my radar.  As I'm getting ready to release the code, I see this still needs addressing.

Re question 1:  so this PR closes a connection that what was previously unclosed yes?  In that case I would not expect any side effects for any code that depends on the fcrepo java client.  Is that your understanding?  Regarding telling users knowing about this `close` feature,  I would just add  a note to the readme.

Re question 2:  I'm not sure that this matters.  Am I missing something?

**Review by dbernstein — APPROVED** [review]

**Comment by tomcbe** [comment]
> @tomcbe : sorry - this PR fell off my radar. As I'm getting ready to release the code, I see this still needs addressing.
> 
> Re question 1: so this PR closes a connection that what was previously unclosed (unless someone supplied their own httpclient) yes? In that case I would not expect any side effects for any code that depends on the fcrepo java client. Is that your understanding? Regarding telling users knowing about this `close` feature, I would just add a note to the readme.

Well in order to actually close the underlying HttpClient instance users should either use FcrepoClient in a try-with-resources block or call FcrepoClient.close() explicitly once they don't need it anymore.

> 
> Re question 2: I'm not sure that this matters. Am I missing something?

I guess it shouldn't have any effects on the outcome of the tests, but would at least be best practice to also manage resources properly in integration tests.

**Comment by tomcbe** [comment]
Hi @dbernstein 
The integration tests needed a refactoring after all. Before that the normally were just hanging at some point because non of the FcrepoResponse objects where properly closed.

Could you please take another look?

File: src/test/java/org/fcrepo/client/integration/FcrepoAuthenticationIT.java [hunk 1]
```diff
@@ -29,109 +35,207 @@
...
 
     @Ignore("Pending alignment with WebAC in FCREPO-2952")
     @Test
     public void testUnAuthUserCannotGet()
             throws Exception {
-        final FcrepoResponse response = client.get(new URI(serverAddress)).perform();
-        final int status = response.getStatusCode();
-        assertEquals("Unauthenticated user should be forbidden!", FORBIDDEN
-                .getStatusCode(), status);
+        try (final FcrepoResponse response = unauthClient.get(new URI(SERVER_ADDRESS)).perform()) {
```
Comment (bbpennel) [inline src/test/java/org/fcrepo/client/integration/FcrepoAuthenticationIT.java:227]: It looks like this method is still ignored, and the ticket it references is "Won't Fix". I wonder if we should re-enable it or delete it. It apparently does not pass

File: src/test/java/org/fcrepo/client/integration/FcrepoClientIT.java [hunk 2]
```diff
@@ -93,29 +102,35 @@ public void testPostBinary() throws Exception {
...
                 .filename(filename)
                 .slug(slug)
-                .perform();
+                .perform()) {
 
-        final String content = IOUtils.toString(response.getBody(), "UTF-8");
-        final int status = response.getStatusCode();
+            final String content = IOUtils.toString(response.getBody(), "UTF-8");
+            final int status = response.getStatusCode();
+            createdRessourceUri = response.getLocation();
```
Comment (bbpennel) [inline src/test/java/org/fcrepo/client/integration/FcrepoClientIT.java:116]:
```suggestion
            createdResourceUri = response.getLocation();
```

File: src/test/java/org/fcrepo/client/integration/FcrepoClientIT.java [hunk 3]
```diff
@@ -570,67 +651,79 @@ public void testGetDisableRedirects() throws Exception {
         final String mimetype = "text/plain";
 
         final URI externalURI = URI.create("http://example.com/");
+        final URI ressourceLocation;
```
Comment (bbpennel) [inline src/test/java/org/fcrepo/client/integration/FcrepoClientIT.java:654]:
```suggestion
        final URI resourceLocation;
```

File: src/test/java/org/fcrepo/client/integration/FcrepoClientIT.java [hunk 4]
```diff
@@ -570,67 +651,79 @@ public void testGetDisableRedirects() throws Exception {
...
     public void testGetWantDigest() throws Exception {
         // Creating a binary for retrieval
         final String mimetype = "text/plain";
         final String bodyContent = "Hello world";
-        final FcrepoResponse response = client.post(new URI(serverAddress))
-                .body(new ByteArrayInputStream(bodyContent.getBytes()), mimetype)
-                .perform();
 
-        final URI url = response.getLocation();
+        final URI createdRessourceUri;
```
Comment (bbpennel) [inline src/test/java/org/fcrepo/client/integration/FcrepoClientIT.java:685]:
```suggestion
        final URI createdResourceUri;
```

File: src/test/java/org/fcrepo/client/integration/FcrepoClientIT.java [hunk 5]
```diff
@@ -570,67 +651,79 @@ public void testGetDisableRedirects() throws Exception {
...
+        }
     }
 
     @Test
     public void testHeadWantDigest() throws Exception {
         // Creating a binary for retrieval
         final String mimetype = "text/plain";
         final String bodyContent = "Hello world";
-        final FcrepoResponse response = client.post(new URI(serverAddress))
+        final URI createdRessourceUri;
```
Comment (bbpennel) [inline src/test/java/org/fcrepo/client/integration/FcrepoClientIT.java:720]:
```suggestion
        final URI createdResourceUri;
```

**Review by bbpennel — CHANGES_REQUESTED** [review]
A few minor questions/suggestions, but over all it looks good to me.

File: src/test/java/org/fcrepo/client/integration/FcrepoAuthenticationIT.java [hunk 6]
```diff
@@ -29,109 +35,207 @@
...
+            final String content = IOUtils.toString(response.getBody(), "UTF-8");
+            final int status = response.getStatusCode();
+            assertEquals("Unauthenticated user should be forbidden! Got content:\n" + content,
+                    FORBIDDEN.getStatusCode(), status);
+        }
     }
 
     @Test
     public void testAuthUserCanPatch() throws Exception {
+        try (final InputStream body = new ByteArrayInputStream(rdfTtl.getBytes())) {
```
[the remark this answers is not in the document]
Comment (tomcbe) [reply src/test/java/org/fcrepo/client/integration/FcrepoAuthenticationIT.java]: Thanks for the hint. I removed the try-with-resources-blocks for the ByteArrayInputstreams in with the last commit.

File: src/test/java/org/fcrepo/client/integration/FcrepoClientIT.java (hunk 2 above)
Comment (tomcbe) [reply src/test/java/org/fcrepo/client/integration/FcrepoClientIT.java]: Fixed locally

File: src/test/java/org/fcrepo/client/integration/FcrepoClientIT.java (hunk 3 above)
Comment (tomcbe) [reply src/test/java/org/fcrepo/client/integration/FcrepoClientIT.java]: Fixed locally

File: src/test/java/org/fcrepo/client/integration/FcrepoClientIT.java (hunk 4 above)
Comment (tomcbe) [reply src/test/java/org/fcrepo/client/integration/FcrepoClientIT.java]: Fixed locally

File: src/test/java/org/fcrepo/client/integration/FcrepoClientIT.java (hunk 5 above)
Comment (tomcbe) [reply src/test/java/org/fcrepo/client/integration/FcrepoClientIT.java]: Fixed locally

**Review by bbpennel — APPROVED** [review]

**Pull request merged**